lib/datasource/maven.js

```diff
@@ -0,0 +1,127 @@
+
+module.exports = {
+  getPkgReleases,
+  resetCache,
+};
+
+let map = new Map();
+
+function resetCache() {
+  map = new Map();
+}
+
+async function getPkgReleases(purl, config) {
+  const versions = [];
+  const dependency = getDependencyParts(purl);
+  const repositories = getRepositories(purl);
+  logger.debug(
+    `Found ${repositories.length} repositories for ${dependency.display}`
+  );
+  for (let i = 0; i < repositories.length; i++) {
+    const repoUrl = repositories[i];
+    logger.debug(
+      `Looking up ${dependency.display} in repository #${i} - ${repoUrl}`
+    );
+    const mavenMetadata = await downloadMavenMetadata(dependency, repoUrl);
+    if (mavenMetadata) {
+      const newVersions = extractVersions(mavenMetadata).filter(
+        version => !versions.includes(version)
+      );
+      versions.push(...newVersions);
+      logger.debug(
+        `Found ${newVersions.length} new versions for ${
+          dependency.display
+        } in repository ${repoUrl}`
+      );
+    }
+  }
+
+  if (versions.length === 0) {
+    logger.warn(
+      `No versions found for ${dependency.display} in ${
+        repositories.length
+      } repositories`
+    );
+    return null;
+  }
+  logger.debug(`Found ${versions.length} versions for ${dependency.display}`);
+
+  return {
+    ...dependency,
+    releases: versions.map(v => ({ version: v })),
+  };
+}
+
+function getDependencyParts(purl) {
+  return {
+    display: `${purl.namespace}:${purl.name}`,
+    group: purl.namespace,
+    name: purl.name,
+    version: purl.version,
+    dependencyUrl: purl.namespace.replace(/\./g, '/') + `/${purl.name}`,
+  };
+}
+
+function getRepositories(purl) {
+  return purl.qualifiers.repository_url.split(',');
+}
+
+async function downloadMavenMetadata(dependency, repoUrl) {
+  const pkgUrl = new url.URL(
+    `${dependency.dependencyUrl}/maven-metadata.xml`,
+    repoUrl
+  );
+  let mavenMetadata;
+  switch (pkgUrl.protocol) {
+    case 'file:':
+      mavenMetadata = await downloadFileProtocol(pkgUrl);
+      break;
+    case 'http:':
+    case 'https:':
+      mavenMetadata = await downloadHttpProtocol(pkgUrl);
+      break;
+  }
+  if (!mavenMetadata) {
+    logger.debug(`${dependency.display} not found in repository ${repoUrl}`);
+  }
+  return mavenMetadata;
+}
+
+function extractVersions(mavenMetadata) {
+  const parser = new DOMParser();
+  const xmlDoc = parser.parseFromString(mavenMetadata, 'text/xml');
+  const versionsTags = xmlDoc.getElementsByTagName('versions')[0].children;
+  const versions = [];
+  for (let i = 0; i < versionsTags.length; i++) {
+    versions.push(versionsTags[i].firstChild.nodeValue);
+  }
+  return versions;
+}
+
+async function downloadFileProtocol(pkgUrl) {
+  const pkgPath = pkgUrl.toString().replace('file://', '');
+  if (!(await fs.exists(pkgPath))) {
+    return null;
+  }
+  return await fs.readFile(pkgPath, 'utf8');
+}
+
+async function downloadHttpProtocol(pkgUrl) {
+  let raw;
+  try {
```

Comment:
not sure which one is the best approach for sending proper, secure get requests. I'm doing some tests to see if this checks are enough

Reply:
What type of "secure" do you mean? Stopping users from manipulating us from querying internal IP addresses?

Reply:
Yeah, not a "secure" maybe is not the best word choice. I mean, if I'm handling correctly all the uses cases: error codes, timeouts, etc? I've seen different ways to do the requests in datasources some more complex than others. I wanted to start with something simple but enough to handle the most common errors.

Reply:
The most important thing is detecting temporary errors in registries (e.g. 429, 5xx) and aborting with the thrown error 'registry-failure'. We do this because otherwise if a registry goes down every few days it would mean that open PRs get autoclosed. The hardest part though is that we can't do it for *every* registry, because sometimes users have custom ones that they know aren't working or authenticated and we don't want Renovate to abort every time and essentially stop working.

So:
- throw error and abort if the main registry fails (e.g. Maven Central?)
- ignore errors from other registries

Reply:
I see, will renovate autoclose a pr for a given dependency if the datasource for whatever reason returns null for that dependency?
So if I throw 'registry-failure' the entire process will be cancelled so renovate won't autoclose any pr?

We can start with your approach, but there are some special cases. For example, it's considered a good practice having your own in-house maven repository, (https://maven.apache.org/repository-management.html) specially with big companies, so you don't smash the public servers.
In this case, your main repository will be your in-house repository. 
I guess we could have some kind of configuration to define repositories that you want to skip, or even define a main repository that make renovate to bail immediately if that repository is down.